proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,617 @@
+  The declaration import namespace brings into availability the subset of the hierarchy of module names rooted in the package "foo-1.3", at the position ``Data.Foo``
+  
+* 2013 de Castro Lopo, qualified exports: `<https://wiki.haskell.org/GHC/QualifiedModuleExport>`_
+  ``qualified module T`` in export list and is essentially a subset of this proposal.
+
+* A worthwhile counter-proposal can be found at `#295`_. That proposal re-casts modules as
+  entities proper; this one, in contrast, continues the historical treatment of modules
+  simply as qualifications to identifiers. In my view, `#295`_ needs to do too much work
+  to keep backward compatibility due to its more fundamental approach. Given a fresh start,
+  I would likely prefer something like `#295`_ than what I have written here, but we're not
+  making a fresh start.
+  
+Alternatives
+------------
+
+Beyond the `Related Work`_, there is wiggle room within this proposal for alternatives.
+
+A. Drop specification point (7) making implicit modules for each type declaration. It's
+   not required by the rest of the proposal. I like it, though.
+
+B. Drop the possibility of the ``import`` keyword in the ``module`` import
+   item. Users can always just write one more ``import module`` statement to
+   unqualify the module's contents. Yet, to me, it just seems more ergonomic
+   to get to the chase right away here.
+
+C. Drop point (8) allowing extension of existing local modules. That was added so that
+   definitions could be mixed in with, e.g., constructors of an implicit local module
+   surrounding a type declaration. But it also seems like a nice way of flexibly mixing
+   modules together. It's an inessential feature.
+
+D. This proposal does not allow the export of a qualified local module such that
+   importers get the identifiers unqualified. We could imagine a new export item
+   ``import module M`` that exports all identifiers in scope with a ``M.`` prefix
+   unqualified. I don't find this feature necessary, but it would fit with the
+   rest of this proposal.
+
+E. Drop the possibility of anonymous local modules; that is, remove ``_`` as a possibility
+   of ``modname``. However, I like the idea of anonymous modules, as they serve to reduce
+   the scope of some declarations without bothering the programmer to write a name for
+   the module.
+
+F. Disallow local modules to be mutually recursive. The current proposal says (point 4)
+   that the local module system affects scoping only. However, we could instead declare
+   that a mutual-dependency strongly-connected component (SCC) cannot include definitions
+   in more than one module. This would disable mutual recursion between modules, but open
+   the possibility of using local modules to explicitly stage compilation. This might
+   allow, for example, the definition of a function and its usage in a Template Haskell
+   splice in the same file (as long as they were in different local modules).
```

Comment:
This does raise the question, though, how do local module interact with Template Haskell's ordering restriction?

Usually when there is a template Haskell slice, a file is split so that toplevel bindings before the slice cannot refer to toplevel bindings after the slice. What happens when there is a slice in the middle of a local module? This probably needs to be specified in the effects and interaction section.